SonarQube.Client/Models/ServerSentEvents/SSEStreamReader.cs

```diff
@@ -0,0 +1,98 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Threading;
+using System;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using SonarQube.Client.Models.ServerSentEvents.ClientContract;
+using SonarQube.Client.Models.ServerSentEvents.ServerContract;
+using System.Collections.Generic;
+using System.Threading.Channels;
+using SonarQube.Client.Logging;
+
+namespace SonarQube.Client.Models.ServerSentEvents
+{
+    public interface ISSEStreamReader
+    {
+        /// <summary>
+        /// Will block the calling thread until an event exists or the connection is closed.
+        /// Can throw an exception if the event is not a valid <see cref="IServerEvent"/>
+        /// </summary>
+        Task<IServerEvent> GetNextEventOrNullAsync();
+    }
+
+    /// <summary>
+    /// Returns <see cref="IServerEvent"/> deserialized from <see cref="ISqServerEvent"/>
+    /// Code on the java side: https://github.com/SonarSource/sonarlint-core/blob/4f34c7c844b12e331a61c63ad7105acac41d2efd/server-api/src/main/java/org/sonarsource/sonarlint/core/serverapi/push/PushApi.java
+    /// </summary>
+    internal class SSEStreamReader : ISSEStreamReader
+    {
+        private readonly ChannelReader<ISqServerEvent> sqEventsChannel;
+        private readonly CancellationToken cancellationToken;
+        private readonly ILogger logger;
+
+        private readonly IDictionary<string, Type> eventConverters = new Dictionary<string, Type>
+        {
+            {"IssueChanged", typeof(IssueChangedServerEvent)},
+            // todo: support later
+            // {"TaintVulnerabilityClosed", typeof(TaintVulnerabilityClosedServerEvent)},
+            // {"TaintVulnerabilityRaised", typeof(TaintVulnerabilityRaisedServerEvent)}
+        };
+
+        public SSEStreamReader(ChannelReader<ISqServerEvent> sqEventsChannel, CancellationToken cancellationToken, ILogger logger)
+        {
+            this.cancellationToken = cancellationToken;
+            this.logger = logger;
+            this.sqEventsChannel = sqEventsChannel;
+        }
+
+        public async Task<IServerEvent> GetNextEventOrNullAsync()
+        {
+            var sqEvent = await sqEventsChannel.ReadAsync(cancellationToken);
+
+            if (sqEvent == null)
+            {
+                return null;
+            }
+
+            if (!eventConverters.ContainsKey(sqEvent.Type))
+            {
+                return null;
+            }
+
+            try
+            {
+                var deserializedEvent = JsonConvert.DeserializeObject(sqEvent.Data, eventConverters[sqEvent.Type]);
+
+                return (IServerEvent) deserializedEvent;
+            }
+            catch (Exception ex)
+            {
+                logger.Debug("[SSEStreamReader] Failed to deserialize sq event." +
```

Comment:
General question: have we considered using a logging library? Some definetely can render contexts for messages (like the `[SSEStreamReader]`)